app/celery/signals.py

```diff
@@ -0,0 +1,103 @@
+import inspect
+import asyncio
+from celery.signals import (
+    task_received, task_internal_error, task_revoked,
+    task_success, task_failure, task_retry, task_prerun
+)
+from celery_singleton.util import generate_lock
+from celery_singleton.config import Config
+from celery_singleton.backends import get_backend
+from app.celery.worker import celery_app
+from app.database import SessionLocal
+from app.services import WebsocketManagerService, JobStatusService, InstructorService
+from app.models.job_status import JobStatusModel, JobStatus
+from app.schemas import JobSchema, WebsocketJobStatusMessage
+
+def handle_job_status_update(job_id: str, job_type: str | None, job_status: JobStatus) -> None:
+    """ Store a job status update in the database and dispatch a websocket event. """
+    async def _handle_job_status_update():
+        with SessionLocal() as session:
+            job_status_model = JobStatusService(session).create_job_status_update(job_id, job_type, job_status)
```

Comment:
Why not have the `JobStatusService` create the db session itself if the lifecycle of the service is within a specific method? I think the constructor of JobStatusService should accept a session as a parameter but default the value to `None`, and if it's `None` then it creates the session for itself. I think that's a common enough occurrence that it makes sense and simplifies creating the service. I think all services we have should use that pattern.

Reply:
Services should not be creating database sessions ever. They are intended to be stateless.

Reply:
That's fair, but then shouldn't this function be stateless as well? We should be passing in the session from the caller instead of creating it here.

Reply:
I'm not sure what you mean. Why would that imply this function should also be stateless?

Reply:
Actually better yet, I don't think we should be storing the status of a job in the db at all. I think we should be storing some sort of state of jobs in an in-memory store like redis.

Reply:
I guess we could, like with an event sourcing stream, but I did think about it a little when implementing this. I don't really see any decisive benefits to it. Sure, we don't necessarily need perpetual persistence of these statuses, but it's not as if it's particularly harmful, so not much of a benefit there. It would be theoretically offer faster querying, but that really doesn't matter at all in this context. We're also not storing simple key-values here, so I'm not even sure using redis streams to implement event sourcing would be viable (which would be the only realistic reason to be using redis for this) since I don't believe they support hashsets.

Reply:
Just to clarify my position here, I don't think it's a bad idea. I think there could be a lot of utility in using Redis throughout our infrastructure to a greater extent than solely to support the job queue and pubsub functionality, so I'm not really opposed to building out the functionality to use Redis differently in areas like this.

But, I do think when we deviate from the primary relational database, we should have good reason to do so and be able to justify that design decision, since things like this introduce more complexity within the project and that's something I think we should seek to avoid.